File: jumpgame1.py
# beginning from position 0, we explore all the paths from the positions reachable from the current.
# We will stop when we reach the end of the array.
# Time complexity - O(n^n)
# Space complexity - O(2n) --we maintain visited array and a stack which can  
# did this code run on leetcode? - No (TLE)
# can also be solved using recursion

from collections import deque

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        if n==1:
            return True
        
        # add the element at nums position 0 in the stack.
        stack = deque([0])
        # visited array
        visited = [False for _ in range(n)] 
        
        while stack:
            top = stack.pop()
            
            # beginpos = min(n-1, nums[top])-top
            
            for i in range(min(n-1,nums[top])-top, -1, -1):
                if top+i == n-1: return True # if it reaches the end.
                # check if the node is not already processed.
                if not visited[top+i]:
                    visited[top+i] = True
                    stack.append(top+i)
            
        return False # if the last position is not reachable.
        
        
# set the destination to the last index of array
# begin traversing from the second last index to the first.
# if the destination is reachable from the new index, set the destination to this index.
# Time complexity - O(n)
# Space complexity - O(1)
# Does this code run on leetcode? - yes
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)

        destination = n-1 
        for i in range(n-2, -1, -1):
            if destination <= nums[i] + i:
                destination = i  # update the destination if the old destination is reachable from the current index.
        return destination == 0

    
# beginning from the top.
# check what is the max reachable node from the beginning of the array
# return True, if it can reach the end. 
# Time complexity - O(n)
# Space complexity - O(1)
# Does this code run on leetcode? - yes
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        
        maxreachable = nums[0] 
        for i in range(1, n):
            if maxreachable >= i:
                maxreachable = max(maxreachable, i+nums[i])
        
        return maxreachable >= n-1

        

File: jumpgame1.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
set reachable = 0
loop over nums
if i > reachable return false (i cannot be more than max reachable from previous elements)
reachable = max(reachable, nums[i]+i)
if we manage to get out of loop then return true
*/
package main

import "fmt"

func canJump(nums []int) bool {
	if len(nums) < 2 {
		return true
	}
	n := len(nums)
	/*
	   //greedy start from last check if we can reach or cross dest from indices before it
	   // if so then reduce dest, if dest reaches 0 then we can go from start to last

	   dest := n - 1
	   for i := n -2; i >=0;i-- {
	       if nums[i] + i >= dest {
	           dest = i
	       }
	   }
	   if dest == 0 {
	       return true
	   } else {
	       return false
	   } */
	//greedy start at beginning, check at each index if we are able to touch reachable, if i > reachable then it means max reachable is less than i so its unreachable with max at each location
	reachable := 0
	for i := 0; i < n; i++ {
		if i > reachable {
			return false
		}
		reachable = max(reachable, nums[i]+i)
	}
	return true
	//return dfs(nums, 0)
}

func dfs(nums []int, ind int) bool {
	//base
	if ind == len(nums)-1 {
		return true
	}
	//logic //if entry is 0 then return false as not possible to move forward
	for i := 1; i <= nums[ind]; i++ {
		if dfs(nums, ind+i) {
			return true
		}
	}
	return false
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func MainCanJump() {
	fmt.Println(canJump([]int{3, 2, 1, 0, 4})) //false
}
